// tdesign-vue-next/es/link/props.mjs
/**
 * tdesign v1.9.4
 * (c) 2024 tdesign
 * @license MIT
 */

var props = {
  content: {
    type: [String, Function]
  },
  "default": {
    type: [String, Function]
  },
  disabled: {
    type: Boolean,
    "default": void 0
  },
  download: {
    type: [String, Boolean]
  },
  hover: {
    type: String,
    "default": "underline",
    validator: function validator(val) {
      if (!val) return true;
      return ["color", "underline"].includes(val);
    }
  },
  href: {
    type: String,
    "default": ""
  },
  prefixIcon: {
    type: Function
  },
  size: {
    type: String,
    "default": "medium",
    validator: function validator(val) {
      if (!val) return true;
      return ["small", "medium", "large"].includes(val);
    }
  },
  suffixIcon: {
    type: Function
  },
  target: {
    type: String,
    "default": ""
  },
  theme: {
    type: String,
    "default": "default",
    validator: function validator(val) {
      if (!val) return true;
      return ["default", "primary", "danger", "warning", "success"].includes(val);
    }
  },
  underline: Boolean,
  onClick: Function
};

export { props as default };
