// tdesign-vue-next/cjs/color-picker/const.js
/**
 * tdesign v1.9.4
 * (c) 2024 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _common_js_colorPicker_constants = require('../_common/js/color-picker/constants.js');

var FORMATS = ["HEX", "RGB", "HSL", "HSV", "CMYK", "CSS"];

exports.COLOR_MODES = _common_js_colorPicker_constants.COLOR_MODES;
exports.DEFAULT_COLOR = _common_js_colorPicker_constants.DEFAULT_COLOR;
exports.DEFAULT_LINEAR_GRADIENT = _common_js_colorPicker_constants.DEFAULT_LINEAR_GRADIENT;
exports.DEFAULT_SYSTEM_SWATCH_COLORS = _common_js_colorPicker_constants.DEFAULT_SYSTEM_SWATCH_COLORS;
exports.GRADIENT_SLIDER_DEFAULT_WIDTH = _common_js_colorPicker_constants.GRADIENT_SLIDER_DEFAULT_WIDTH;
exports.SATURATION_PANEL_DEFAULT_HEIGHT = _common_js_colorPicker_constants.SATURATION_PANEL_DEFAULT_HEIGHT;
exports.SATURATION_PANEL_DEFAULT_WIDTH = _common_js_colorPicker_constants.SATURATION_PANEL_DEFAULT_WIDTH;
exports.SLIDER_DEFAULT_WIDTH = _common_js_colorPicker_constants.SLIDER_DEFAULT_WIDTH;
exports.TD_COLOR_USED_COLORS_MAX_SIZE = _common_js_colorPicker_constants.TD_COLOR_USED_COLORS_MAX_SIZE;
exports.FORMATS = FORMATS;
