// tdesign-vue-next/es/color-picker/color-picker-panel.mjs
/**
 * tdesign v1.9.4
 * (c) 2024 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-2dead3b8.mjs';
import { defineComponent, computed, createVNode, mergeProps } from 'vue';
import props from './props.mjs';
import ColorPanel from './panel/index.mjs';
import { usePrefixClass } from '../hooks/useConfig.mjs';
import '../_chunks/dep-ad0f5c12.mjs';
import { _ as _arrayMap } from '../_chunks/dep-056e3d49.mjs';
import { _ as _baseIteratee } from '../_chunks/dep-600393b8.mjs';
import { _ as _basePickBy } from '../_chunks/dep-85cd63c4.mjs';
import { _ as _getAllKeysIn } from '../_chunks/dep-12db5b5b.mjs';
import '../_chunks/dep-8a6a0750.mjs';
import '../_chunks/dep-83cdf63a.mjs';
import '../_chunks/dep-04fa0fa7.mjs';
import '../_chunks/dep-f29ee43a.mjs';
import './const.mjs';
import '../_common/js/color-picker/constants.mjs';
import './panel/header.mjs';
import '../radio/index.mjs';
import '../radio/radio.mjs';
import '../utils/helper.mjs';
import '../_chunks/dep-28f56d24.mjs';
import '../_chunks/dep-084c0ba6.mjs';
import '../_chunks/dep-07b6df34.mjs';
import '../_chunks/dep-441fa01f.mjs';
import '../_chunks/dep-7d8aa418.mjs';
import '../_chunks/dep-fb07e55c.mjs';
import '../_chunks/dep-660119a6.mjs';
import '../_chunks/dep-4ef63f67.mjs';
import '../_chunks/dep-844e471b.mjs';
import '../_chunks/dep-de0d9668.mjs';
import '../_chunks/dep-c23362d5.mjs';
import '../_chunks/dep-d5cc8b5d.mjs';
import '../_chunks/dep-f5bde8c1.mjs';
import '../radio/props.mjs';
import '../radio/constants.mjs';
import '../form/hooks.mjs';
import '../hooks/useVModel.mjs';
import '../_chunks/dep-dab87894.mjs';
import '../hooks/tnode.mjs';
import '../_chunks/dep-706c37a3.mjs';
import '../_chunks/dep-a053ff3d.mjs';
import '../_chunks/dep-65c48aa3.mjs';
import '../utils/render-tnode.mjs';
import '../_chunks/dep-a56119c3.mjs';
import '../_chunks/dep-f8f5e5fe.mjs';
import '../_chunks/dep-1a2263f9.mjs';
import '../_chunks/dep-0126dffd.mjs';
import '../_chunks/dep-82bdd07b.mjs';
import '../_chunks/dep-64862a5a.mjs';
import '../_chunks/dep-48235a7f.mjs';
import '../_chunks/dep-5c0cf5f9.mjs';
import '../_chunks/dep-c6ed1157.mjs';
import '../config-provider/useConfig.mjs';
import '../_chunks/dep-70894ccf.mjs';
import '../_chunks/dep-88e27495.mjs';
import '../_chunks/dep-b1040349.mjs';
import '../_chunks/dep-203afd59.mjs';
import '../_chunks/dep-bfe0dcb6.mjs';
import '../_chunks/dep-e8c6972d.mjs';
import '../_chunks/dep-b15f1e4a.mjs';
import '../_chunks/dep-9b1e7043.mjs';
import '../_chunks/dep-5b74d046.mjs';
import '../_chunks/dep-551a93fd.mjs';
import '../_chunks/dep-672a56f9.mjs';
import '../_chunks/dep-8c0df173.mjs';
import '../_chunks/dep-c3f9042a.mjs';
import '../_common/js/global-config/default-config.mjs';
import '../_common/js/global-config/locale/zh_CN.mjs';
import '../_chunks/dep-841260ab.mjs';
import '../_chunks/dep-22c04ae0.mjs';
import '../_chunks/dep-886b81c5.mjs';
import '../config-provider/type.mjs';
import '../_chunks/dep-45a70a53.mjs';
import '../_chunks/dep-8e0cf610.mjs';
import '../radio/group.mjs';
import '../_chunks/dep-8a25f93c.mjs';
import '../_chunks/dep-252b88a1.mjs';
import '../_chunks/dep-eb82bf9c.mjs';
import '../_chunks/dep-6cd78bb0.mjs';
import '../_chunks/dep-786a7628.mjs';
import '../radio/radio-group-props.mjs';
import '../radio/useKeyboard.mjs';
import '../utils/dom.mjs';
import '../utils/easing.mjs';
import '../_common/js/common.mjs';
import '../watermark/hooks.mjs';
import '../hooks/useResizeObserver.mjs';
import '../radio/radio-button.mjs';
import '../utils/withInstall.mjs';
import './style/css.mjs';
import '../radio/type.mjs';
import './hooks.mjs';
import './panel/linear-gradient.mjs';
import './utils/index.mjs';
import '../_common/js/color-picker/index.mjs';
import '../_common/js/color-picker/cmyk.mjs';
import '../_common/js/color-picker/color.mjs';
import '../_chunks/dep-d470848d.mjs';
import '../_chunks/dep-a116202e.mjs';
import '../_common/js/color-picker/gradient.mjs';
import '../_common/js/color-picker/draggable.mjs';
import '../input-number/index.mjs';
import '../input-number/input-number.mjs';
import 'tdesign-icons-vue-next';
import '../button/index.mjs';
import '../button/button.mjs';
import '../loading/index.mjs';
import '../loading/loading.mjs';
import '../loading/icon/gradient.mjs';
import '../_common/js/loading/circle-adapter.mjs';
import '../_common/js/utils/set-style.mjs';
import '../_common/js/utils/helper.mjs';
import '../loading/props.mjs';
import '../hooks/useTeleport.mjs';
import '../loading/type.mjs';
import '../loading/plugin.mjs';
import '../_chunks/dep-7c4d232a.mjs';
import '../_chunks/dep-1cfe79e3.mjs';
import '../_chunks/dep-2a455a8f.mjs';
import '../_chunks/dep-87670918.mjs';
import '../_chunks/dep-ee7aa1c5.mjs';
import '../_chunks/dep-6f363254.mjs';
import '../_chunks/dep-09c30dd8.mjs';
import '../button/props.mjs';
import '../hooks/useRipple.mjs';
import '../hooks/useKeepAnimation.mjs';
import '../utils/set-style.mjs';
import '../hooks/useDisabled.mjs';
import '../_chunks/dep-fb48187a.mjs';
import '../button/type.mjs';
import '../input/index.mjs';
import '../input/input.mjs';
import '../input/props.mjs';
import '../hooks/useGlobalIcon.mjs';
import '../input/useInput.mjs';
import '../form/const.mjs';
import '../hooks/index.mjs';
import '../hooks/icon.mjs';
import '../hooks/slot.mjs';
import '../hooks/useCommonClassName.mjs';
import '../hooks/useDefaultValue.mjs';
import '../hooks/useDestroyOnClose.mjs';
import '../hooks/useVirtualScroll.mjs';
import '../hooks/useImagePreviewUrl.mjs';
import '../_common/js/upload/utils.mjs';
import '../_common/js/log/log.mjs';
import '../input/useLengthLimit.mjs';
import '../_common/js/log/index.mjs';
import '../input/useInputEventHandler.mjs';
import '../input/useInputWidth.mjs';
import '../input/input-group.mjs';
import '../input/input-group-props.mjs';
import '../input/type.mjs';
import '../input-number/props.mjs';
import '../input-number/useInputNumber.mjs';
import '../_common/js/input-number/number.mjs';
import '../_common/js/input-number/large-number.mjs';
import '../input-number/type.mjs';
import './panel/base-props.mjs';
import './panel/saturation.mjs';
import './panel/hue.mjs';
import './panel/slider.mjs';
import './panel/alpha.mjs';
import './panel/format/index.mjs';
import '../select/index.mjs';
import '../select/select.mjs';
import '../_chunks/dep-a5f8156f.mjs';
import '../_chunks/dep-f5c953e5.mjs';
import '../common-components/fake-arrow.mjs';
import '../select-input/index.mjs';
import '../select-input/select-input.mjs';
import '../popup/index.mjs';
import '../popup/popup.mjs';
import '@popperjs/core';
import '../popup/container.mjs';
import '../popup/props.mjs';
import '../popup/type.mjs';
import '../select-input/props.mjs';
import '../select-input/useSingle.mjs';
import '../select-input/useMultiple.mjs';
import '../tag-input/index.mjs';
import '../tag-input/tag-input.mjs';
import '../tag-input/props.mjs';
import '../tag-input/hooks/useTagScroll.mjs';
import '../tag-input/useTagList.mjs';
import '../tag/index.mjs';
import '../tag/tag.mjs';
import '../tag/props.mjs';
import '../tag/check-tag.mjs';
import '../tag/check-tag-props.mjs';
import '../tag/check-tag-group.mjs';
import '../tag/check-tag-group-props.mjs';
import '../tag/type.mjs';
import '../tag-input/hooks/useHover.mjs';
import '../tag-input/hooks/useDragSorter.mjs';
import '../tag-input/type.mjs';
import '../select-input/useOverlayInnerStyle.mjs';
import '../_chunks/dep-f6f516d9.mjs';
import '../_chunks/dep-a1a5cc96.mjs';
import '../select-input/type.mjs';
import '../select/select-panel.mjs';
import '../_chunks/dep-281dbb92.mjs';
import '../_chunks/dep-c0c2d9c7.mjs';
import '../select/option.mjs';
import '../select/option-props.mjs';
import '../checkbox/index.mjs';
import '../checkbox/checkbox.mjs';
import '../checkbox/props.mjs';
import '../checkbox/constants.mjs';
import '../checkbox/hooks/useCheckboxLazyLoad.mjs';
import '../_common/js/utils/observe.mjs';
import '../checkbox/hooks/useKeyboardEvent.mjs';
import '../checkbox/group.mjs';
import '../checkbox/checkbox-group-props.mjs';
import '../checkbox/type.mjs';
import '../select/helper.mjs';
import '../select/optionGroup.mjs';
import '../select/option-group-props.mjs';
import '../select/props.mjs';
import '../select/hooks/usePanelVirtualScroll.mjs';
import '../select/hooks/useSelectOptions.mjs';
import '../select/hooks/useKeyboardControl.mjs';
import '../select/type.mjs';
import './panel/format/inputs.mjs';
import './panel/format/config.mjs';
import './panel/swatches.mjs';

var arrayMap = _arrayMap,
  baseIteratee = _baseIteratee,
  basePickBy = _basePickBy,
  getAllKeysIn = _getAllKeysIn;

/**
 * Creates an object composed of the `object` properties `predicate` returns
 * truthy for. The predicate is invoked with two arguments: (value, key).
 *
 * @static
 * @memberOf _
 * @since 4.0.0
 * @category Object
 * @param {Object} object The source object.
 * @param {Function} [predicate=_.identity] The function invoked per property.
 * @returns {Object} Returns the new object.
 * @example
 *
 * var object = { 'a': 1, 'b': '2', 'c': 3 };
 *
 * _.pickBy(object, _.isNumber);
 * // => { 'a': 1, 'c': 3 }
 */
function pickBy(object, predicate) {
  if (object == null) {
    return {};
  }
  var props = arrayMap(getAllKeysIn(object), function (prop) {
    return [prop];
  });
  predicate = baseIteratee(predicate);
  return basePickBy(object, props, function (value, path) {
    return predicate(value, path[0]);
  });
}
var pickBy_1 = pickBy;

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var _ColorPickerPanel = defineComponent({
  name: "TColorPickerPanel",
  inheritAttrs: false,
  props: _objectSpread({}, props),
  setup: function setup(props2, _ref) {
    var attrs = _ref.attrs;
    var newProps = computed(function () {
      return pickBy_1(_objectSpread(_objectSpread({}, props2), attrs), function (v) {
        return v !== void 0;
      });
    });
    var prefix = usePrefixClass();
    return function () {
      return createVNode(ColorPanel, mergeProps(newProps.value, {
        "popupProps": null,
        "close-btn": false,
        "class": "".concat(prefix.value, "-is-inline")
      }), null);
    };
  }
});

export { _ColorPickerPanel as default };
