// tdesign-vue-next/es/color-picker/props.mjs
/**
 * tdesign v1.9.4
 * (c) 2024 tdesign
 * @license MIT
 */

var props = {
  clearable: Boolean,
  closeBtn: {
    type: [String, Boolean, Function],
    "default": true
  },
  colorModes: {
    type: Array,
    "default": function _default() {
      return ["monochrome", "linear-gradient"];
    }
  },
  disabled: Boolean,
  enableAlpha: Boolean,
  enableMultipleGradient: {
    type: Boolean,
    "default": true
  },
  format: {
    type: String,
    "default": "RGB",
    validator: function validator(val) {
      if (!val) return true;
      return ["RGB", "RGBA", "HSL", "HSLA", "HSB", "HSV", "HSVA", "HEX", "CMYK", "CSS"].includes(val);
    }
  },
  inputProps: {
    type: Object
  },
  multiple: Boolean,
  popupProps: {
    type: Object
  },
  recentColors: {
    type: Array,
    "default": void 0
  },
  defaultRecentColors: {
    type: Array,
    "default": function _default() {
      return [];
    }
  },
  selectInputProps: {
    type: Object
  },
  showPrimaryColorPreview: {
    type: Boolean,
    "default": true
  },
  size: {
    type: String,
    "default": "medium",
    validator: function validator(val) {
      if (!val) return true;
      return ["small", "medium", "large"].includes(val);
    }
  },
  swatchColors: {
    type: Array
  },
  value: {
    type: String,
    "default": void 0
  },
  modelValue: {
    type: String,
    "default": void 0
  },
  defaultValue: {
    type: String,
    "default": ""
  },
  onChange: Function,
  onPaletteBarChange: Function,
  onRecentColorsChange: Function
};

export { props as default };
