// tdesign-vue-next/lib/tag-input/props.js
/**
 * tdesign v1.9.4
 * (c) 2024 tdesign
 * @license MIT
 */

var props = {
  autoWidth: Boolean,
  clearable: Boolean,
  collapsedItems: {
    type: Function
  },
  disabled: Boolean,
  dragSort: Boolean,
  excessTagsDisplayType: {
    type: String,
    "default": "break-line",
    validator: function validator(val) {
      if (!val) return true;
      return ["scroll", "break-line"].includes(val);
    }
  },
  inputProps: {
    type: Object
  },
  inputValue: {
    type: [String, Number],
    "default": void 0
  },
  defaultInputValue: {
    type: [String, Number],
    "default": ""
  },
  label: {
    type: [String, Function]
  },
  max: {
    type: Number
  },
  minCollapsedNum: {
    type: Number,
    "default": 0
  },
  placeholder: {
    type: String,
    "default": void 0
  },
  prefixIcon: {
    type: Function
  },
  readonly: Boolean,
  size: {
    type: String,
    "default": "medium",
    validator: function validator(val) {
      if (!val) return true;
      return ["small", "medium", "large"].includes(val);
    }
  },
  status: {
    type: String,
    validator: function validator(val) {
      if (!val) return true;
      return ["default", "success", "warning", "error"].includes(val);
    }
  },
  suffix: {
    type: [String, Function]
  },
  suffixIcon: {
    type: Function
  },
  tag: {
    type: [String, Function]
  },
  tagProps: {
    type: Object
  },
  tips: {
    type: [String, Function]
  },
  value: {
    type: Array,
    "default": void 0
  },
  modelValue: {
    type: Array,
    "default": void 0
  },
  defaultValue: {
    type: Array,
    "default": function _default() {
      return [];
    }
  },
  valueDisplay: {
    type: [String, Function]
  },
  onBlur: Function,
  onChange: Function,
  onClear: Function,
  onClick: Function,
  onDragSort: Function,
  onEnter: Function,
  onFocus: Function,
  onInputChange: Function,
  onMouseenter: Function,
  onMouseleave: Function,
  onPaste: Function,
  onRemove: Function
};

export { props as default };
