// tdesign-vue-next/cjs/_chunks/dep-6de3a26f.js
/**
 * tdesign v1.9.4
 * (c) 2024 tdesign
 * @license MIT
 */

'use strict';

var _typeof = require('@babel/runtime/helpers/typeof');
var _commonjsHelpers = require('./dep-4980a22e.js');
var dayjs = require('dayjs');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _typeof__default = /*#__PURE__*/_interopDefaultLegacy(_typeof);
var dayjs__default = /*#__PURE__*/_interopDefaultLegacy(dayjs);

var it = {exports: {}};

(function (module, exports) {
  !function (e, o) {
    module.exports = o(dayjs__default["default"]) ;
  }(_commonjsHelpers.commonjsGlobal, function (e) {

    function o(e) {
      return e && "object" == _typeof__default["default"](e) && "default" in e ? e : {
        "default": e
      };
    }
    var t = o(e),
      n = {
        name: "it",
        weekdays: "domenica_lunedì_martedì_mercoledì_giovedì_venerdì_sabato".split("_"),
        weekdaysShort: "dom_lun_mar_mer_gio_ven_sab".split("_"),
        weekdaysMin: "do_lu_ma_me_gi_ve_sa".split("_"),
        months: "gennaio_febbraio_marzo_aprile_maggio_giugno_luglio_agosto_settembre_ottobre_novembre_dicembre".split("_"),
        weekStart: 1,
        monthsShort: "gen_feb_mar_apr_mag_giu_lug_ago_set_ott_nov_dic".split("_"),
        formats: {
          LT: "HH:mm",
          LTS: "HH:mm:ss",
          L: "DD/MM/YYYY",
          LL: "D MMMM YYYY",
          LLL: "D MMMM YYYY HH:mm",
          LLLL: "dddd D MMMM YYYY HH:mm"
        },
        relativeTime: {
          future: "tra %s",
          past: "%s fa",
          s: "qualche secondo",
          m: "un minuto",
          mm: "%d minuti",
          h: "un' ora",
          hh: "%d ore",
          d: "un giorno",
          dd: "%d giorni",
          M: "un mese",
          MM: "%d mesi",
          y: "un anno",
          yy: "%d anni"
        },
        ordinal: function ordinal(e) {
          return e + "º";
        }
      };
    return t["default"].locale(n, null, !0), n;
  });
})(it);
