// tdesign-vue-next/cjs/_chunks/dep-b9072679.js
/**
 * tdesign v1.9.4
 * (c) 2024 tdesign
 * @license MIT
 */

'use strict';

var _typeof = require('@babel/runtime/helpers/typeof');
var _commonjsHelpers = require('./dep-4980a22e.js');
var dayjs = require('dayjs');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _typeof__default = /*#__PURE__*/_interopDefaultLegacy(_typeof);
var dayjs__default = /*#__PURE__*/_interopDefaultLegacy(dayjs);

var ru = {exports: {}};

(function (module, exports) {
  !function (_, t) {
    module.exports = t(dayjs__default["default"]) ;
  }(_commonjsHelpers.commonjsGlobal, function (_) {

    function t(_) {
      return _ && "object" == _typeof__default["default"](_) && "default" in _ ? _ : {
        "default": _
      };
    }
    var e = t(_),
      n = "января_февраля_марта_апреля_мая_июня_июля_августа_сентября_октября_ноября_декабря".split("_"),
      s = "январь_февраль_март_апрель_май_июнь_июль_август_сентябрь_октябрь_ноябрь_декабрь".split("_"),
      r = "янв._февр._мар._апр._мая_июня_июля_авг._сент._окт._нояб._дек.".split("_"),
      o = "янв._февр._март_апр._май_июнь_июль_авг._сент._окт._нояб._дек.".split("_"),
      i = /D[oD]?(\[[^[\]]*\]|\s)+MMMM?/;
    function d(_, t, e) {
      var n, s;
      return "m" === e ? t ? "минута" : "минуту" : _ + " " + (n = +_, s = {
        mm: t ? "минута_минуты_минут" : "минуту_минуты_минут",
        hh: "час_часа_часов",
        dd: "день_дня_дней",
        MM: "месяц_месяца_месяцев",
        yy: "год_года_лет"
      }[e].split("_"), n % 10 == 1 && n % 100 != 11 ? s[0] : n % 10 >= 2 && n % 10 <= 4 && (n % 100 < 10 || n % 100 >= 20) ? s[1] : s[2]);
    }
    var u = function u(_, t) {
      return i.test(t) ? n[_.month()] : s[_.month()];
    };
    u.s = s, u.f = n;
    var a = function a(_, t) {
      return i.test(t) ? r[_.month()] : o[_.month()];
    };
    a.s = o, a.f = r;
    var m = {
      name: "ru",
      weekdays: "воскресенье_понедельник_вторник_среда_четверг_пятница_суббота".split("_"),
      weekdaysShort: "вск_пнд_втр_срд_чтв_птн_сбт".split("_"),
      weekdaysMin: "вс_пн_вт_ср_чт_пт_сб".split("_"),
      months: u,
      monthsShort: a,
      weekStart: 1,
      yearStart: 4,
      formats: {
        LT: "H:mm",
        LTS: "H:mm:ss",
        L: "DD.MM.YYYY",
        LL: "D MMMM YYYY г.",
        LLL: "D MMMM YYYY г., H:mm",
        LLLL: "dddd, D MMMM YYYY г., H:mm"
      },
      relativeTime: {
        future: "через %s",
        past: "%s назад",
        s: "несколько секунд",
        m: d,
        mm: d,
        h: "час",
        hh: d,
        d: "день",
        dd: d,
        M: "месяц",
        MM: d,
        y: "год",
        yy: d
      },
      ordinal: function ordinal(_) {
        return _;
      },
      meridiem: function meridiem(_) {
        return _ < 4 ? "ночи" : _ < 12 ? "утра" : _ < 17 ? "дня" : "вечера";
      }
    };
    return e["default"].locale(m, null, !0), m;
  });
})(ru);
