// tdesign-vue-next/cjs/_chunks/dep-c9bf9615.js
/**
 * tdesign v1.9.4
 * (c) 2024 tdesign
 * @license MIT
 */

'use strict';

var _typeof = require('@babel/runtime/helpers/typeof');
var _commonjsHelpers = require('./dep-4980a22e.js');
var dayjs = require('dayjs');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _typeof__default = /*#__PURE__*/_interopDefaultLegacy(_typeof);
var dayjs__default = /*#__PURE__*/_interopDefaultLegacy(dayjs);

var ko = {exports: {}};

(function (module, exports) {
  !function (e, _) {
    module.exports = _(dayjs__default["default"]) ;
  }(_commonjsHelpers.commonjsGlobal, function (e) {

    function _(e) {
      return e && "object" == _typeof__default["default"](e) && "default" in e ? e : {
        "default": e
      };
    }
    var d = _(e),
      t = {
        name: "ko",
        weekdays: "일요일_월요일_화요일_수요일_목요일_금요일_토요일".split("_"),
        weekdaysShort: "일_월_화_수_목_금_토".split("_"),
        weekdaysMin: "일_월_화_수_목_금_토".split("_"),
        months: "1월_2월_3월_4월_5월_6월_7월_8월_9월_10월_11월_12월".split("_"),
        monthsShort: "1월_2월_3월_4월_5월_6월_7월_8월_9월_10월_11월_12월".split("_"),
        ordinal: function ordinal(e) {
          return e + "일";
        },
        formats: {
          LT: "A h:mm",
          LTS: "A h:mm:ss",
          L: "YYYY.MM.DD.",
          LL: "YYYY년 MMMM D일",
          LLL: "YYYY년 MMMM D일 A h:mm",
          LLLL: "YYYY년 MMMM D일 dddd A h:mm",
          l: "YYYY.MM.DD.",
          ll: "YYYY년 MMMM D일",
          lll: "YYYY년 MMMM D일 A h:mm",
          llll: "YYYY년 MMMM D일 dddd A h:mm"
        },
        meridiem: function meridiem(e) {
          return e < 12 ? "오전" : "오후";
        },
        relativeTime: {
          future: "%s 후",
          past: "%s 전",
          s: "몇 초",
          m: "1분",
          mm: "%d분",
          h: "한 시간",
          hh: "%d시간",
          d: "하루",
          dd: "%d일",
          M: "한 달",
          MM: "%d달",
          y: "일 년",
          yy: "%d년"
        }
      };
    return d["default"].locale(t, null, !0), t;
  });
})(ko);
