// tdesign-vue-next/cjs/_chunks/dep-648bbf41.js
/**
 * tdesign v1.9.4
 * (c) 2024 tdesign
 * @license MIT
 */

'use strict';

var button_index = require('../button/index.js');
var icon_index = require('../icon/index.js');
var link_index = require('../link/index.js');
var grid_index = require('../grid/index.js');
var layout_index = require('../layout/index.js');
var divider_index = require('../divider/index.js');
var space_index = require('../space/index.js');
var descriptions_index = require('../descriptions/index.js');
var affix_index = require('../affix/index.js');
var anchor_index = require('../anchor/index.js');
var breadcrumb_index = require('../breadcrumb/index.js');
var dropdown_index = require('../dropdown/index.js');
var menu_index = require('../menu/index.js');
var pagination_index = require('../pagination/index.js');
var steps_index = require('../steps/index.js');
var tabs_index = require('../tabs/index.js');
var stickyTool_index = require('../sticky-tool/index.js');
var backTop_index = require('../back-top/index.js');
var autoComplete_index = require('../auto-complete/index.js');
var cascader_index = require('../cascader/index.js');
var checkbox_index = require('../checkbox/index.js');
var datePicker_index = require('../date-picker/index.js');
var form_index = require('../form/index.js');
var inputAdornment_index = require('../input-adornment/index.js');
var input_index = require('../input/index.js');
var inputNumber_index = require('../input-number/index.js');
var radio_index = require('../radio/index.js');
var select_index = require('../select/index.js');
var slider_index = require('../slider/index.js');
var switch_index = require('../switch/index.js');
var tagInput_index = require('../tag-input/index.js');
var selectInput_index = require('../select-input/index.js');
var textarea_index = require('../textarea/index.js');
var transfer_index = require('../transfer/index.js');
var timePicker_index = require('../time-picker/index.js');
var treeSelect_index = require('../tree-select/index.js');
var colorPicker_index = require('../color-picker/index.js');
var rangeInput_index = require('../range-input/index.js');
var avatar_index = require('../avatar/index.js');
var badge_index = require('../badge/index.js');
var calendar_index = require('../calendar/index.js');
var card_index = require('../card/index.js');
var comment_index = require('../comment/index.js');
var image_index = require('../image/index.js');
var imageViewer_index = require('../image-viewer/index.js');
var list_index = require('../list/index.js');
var progress_index = require('../progress/index.js');
var skeleton_index = require('../skeleton/index.js');
var statistic_index = require('../statistic/index.js');
var swiper_index = require('../swiper/index.js');
var table_index = require('../table/index.js');
var tag_index = require('../tag/index.js');
var timeline_index = require('../timeline/index.js');
var tooltip_index = require('../tooltip/index.js');
var tree_index = require('../tree/index.js');
var collapse_index = require('../collapse/index.js');
var watermark_index = require('../watermark/index.js');
var rate_index = require('../rate/index.js');
var alert_index = require('../alert/index.js');
var dialog_index = require('../dialog/index.js');
var drawer_index = require('../drawer/index.js');
var guide_index = require('../guide/index.js');
var loading_index = require('../loading/index.js');
var message_index = require('../message/index.js');
var notification_index = require('../notification/index.js');
var popconfirm_index = require('../popconfirm/index.js');
var popup_index = require('../popup/index.js');
var upload_index = require('../upload/index.js');
var configProvider_index = require('../config-provider/index.js');
var dialog_plugin = require('../dialog/plugin.js');
var loading_plugin = require('../loading/plugin.js');
var loading_directive = require('../loading/directive.js');
var message_plugin = require('../message/plugin.js');
var notification_plugin = require('../notification/plugin.js');

var components = /*#__PURE__*/Object.freeze({
	__proto__: null,
	Button: button_index.Button,
	Icon: icon_index.Icon,
	Link: link_index.Link,
	Row: grid_index.Row,
	Col: grid_index.Col,
	Aside: layout_index.Aside,
	Layout: layout_index.Layout,
	Header: layout_index.Header,
	Footer: layout_index.Footer,
	Content: layout_index.Content,
	Divider: divider_index.Divider,
	Space: space_index.Space,
	Descriptions: descriptions_index.Descriptions,
	DescriptionsItem: descriptions_index.DescriptionsItem,
	Affix: affix_index.Affix,
	Anchor: anchor_index.Anchor,
	AnchorItem: anchor_index.AnchorItem,
	AnchorTarget: anchor_index.AnchorTarget,
	Breadcrumb: breadcrumb_index.Breadcrumb,
	BreadcrumbItem: breadcrumb_index.BreadcrumbItem,
	Dropdown: dropdown_index.Dropdown,
	DropdownItem: dropdown_index.DropdownItem,
	DropdownMenu: dropdown_index.DropdownMenu,
	Menu: menu_index.Menu,
	HeadMenu: menu_index.HeadMenu,
	Submenu: menu_index.Submenu,
	MenuItem: menu_index.MenuItem,
	MenuGroup: menu_index.MenuGroup,
	Pagination: pagination_index.Pagination,
	PaginationMini: pagination_index.PaginationMini,
	Steps: steps_index.Steps,
	StepItem: steps_index.StepItem,
	Tabs: tabs_index.Tabs,
	TabPanel: tabs_index.TabPanel,
	StickyItem: stickyTool_index.StickyItem,
	StickyTool: stickyTool_index.StickyTool,
	BackTop: backTop_index.BackTop,
	AutoComplete: autoComplete_index.AutoComplete,
	HighlightOption: autoComplete_index.HighlightOption,
	Cascader: cascader_index.Cascader,
	CascaderPanel: cascader_index.CascaderPanel,
	Checkbox: checkbox_index.Checkbox,
	CheckboxGroup: checkbox_index.CheckboxGroup,
	DatePicker: datePicker_index.DatePicker,
	DatePickerPanel: datePicker_index.DatePickerPanel,
	DateRangePicker: datePicker_index.DateRangePicker,
	DateRangePickerPanel: datePicker_index.DateRangePickerPanel,
	Form: form_index.Form,
	FormItem: form_index.FormItem,
	InputAdornment: inputAdornment_index.InputAdornment,
	Input: input_index.Input,
	InputGroup: input_index.InputGroup,
	InputNumber: inputNumber_index.InputNumber,
	Radio: radio_index.Radio,
	RadioGroup: radio_index.RadioGroup,
	RadioButton: radio_index.RadioButton,
	Select: select_index.Select,
	Option: select_index.Option,
	OptionGroup: select_index.OptionGroup,
	Slider: slider_index.Slider,
	SliderButton: slider_index.SliderButton,
	Switch: switch_index.Switch,
	TagInput: tagInput_index.TagInput,
	SelectInput: selectInput_index.SelectInput,
	Textarea: textarea_index.Textarea,
	Transfer: transfer_index.Transfer,
	TimePicker: timePicker_index.TimePicker,
	TimeRangePicker: timePicker_index.TimeRangePicker,
	TimePickerPanel: timePicker_index.TimePickerPanel,
	TreeSelect: treeSelect_index.TreeSelect,
	ColorPickerPanel: colorPicker_index.ColorPickerPanel,
	ColorPicker: colorPicker_index.ColorPicker,
	RangeInput: rangeInput_index.RangeInput,
	RangeInputPopup: rangeInput_index.RangeInputPopup,
	Avatar: avatar_index.Avatar,
	AvatarGroup: avatar_index.AvatarGroup,
	Badge: badge_index.Badge,
	Calendar: calendar_index.Calendar,
	Card: card_index.Card,
	Comment: comment_index.Comment,
	Image: image_index.Image,
	ImageViewer: imageViewer_index.ImageViewer,
	List: list_index.List,
	ListItem: list_index.ListItem,
	ListItemMeta: list_index.ListItemMeta,
	Progress: progress_index.Progress,
	Skeleton: skeleton_index.Skeleton,
	Statistic: statistic_index.Statistic,
	Swiper: swiper_index.Swiper,
	SwiperItem: swiper_index.SwiperItem,
	BaseTable: table_index.BaseTable,
	PrimaryTable: table_index.PrimaryTable,
	EnhancedTable: table_index.EnhancedTable,
	Table: table_index.Table,
	Tag: tag_index.Tag,
	CheckTag: tag_index.CheckTag,
	CheckTagGroup: tag_index.CheckTagGroup,
	Timeline: timeline_index.Timeline,
	TimelineItem: timeline_index.TimelineItem,
	Tooltip: tooltip_index.Tooltip,
	Tree: tree_index.Tree,
	Collapse: collapse_index.Collapse,
	CollapsePanel: collapse_index.CollapsePanel,
	Watermark: watermark_index.Watermark,
	Rate: rate_index.Rate,
	Alert: alert_index.Alert,
	DialogPlugin: dialog_plugin.DialogPlugin,
	Dialog: dialog_index.Dialog,
	Drawer: drawer_index.Drawer,
	Guide: guide_index.Guide,
	LoadingPlugin: loading_plugin.LoadingPlugin,
	LoadingDirective: loading_directive.vLoading,
	Loading: loading_index.Loading,
	MessagePlugin: message_plugin.MessagePlugin,
	Message: message_index.Message,
	NotifyPlugin: notification_plugin["default"],
	Notification: notification_index.Notification,
	Popconfirm: popconfirm_index.Popconfirm,
	Popup: popup_index.Popup,
	Upload: upload_index.Upload,
	ConfigProvider: configProvider_index.ConfigProvider
});

exports.components = components;
