// tdesign-vue-next/cjs/_common/js/time-picker/const.js
/**
 * tdesign v1.9.4
 * (c) 2024 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var AM = "am";
var PM = "pm";
var MERIDIEM_LIST = [AM, PM];
var TIME_PICKER_EMPTY = [void 0, void 0];
var DEFAULT_STEPS = [1, 1, 1];
var DEFAULT_FORMAT = "HH:mm:ss";
var EMPTY_VALUE = -1;
var EPickerCols = /* @__PURE__ */function (EPickerCols2) {
  EPickerCols2["hour"] = "hour";
  EPickerCols2["minute"] = "minute";
  EPickerCols2["second"] = "second";
  EPickerCols2["milliSecond"] = "millisecond";
  EPickerCols2["meridiem"] = "meridiem";
  return EPickerCols2;
}(EPickerCols || {});
var TIME_FORMAT = /(a\s+|A\s+)?(h+|H+)?:?(m+)?:?(s+)?:?(S+)?(\s+a|\s+A)?/;
var PRE_MERIDIEM_FORMAT = /^(a|A)\s+?[h]{1,2}(:[m]{1,2}(:[s]{1,2})?)?$/;
var POST_MERIDIEM_FORMAT = /^[h]{1,2}(:[m]{1,2}(:[s]{1,2})?)?(\s+(a|A))?$/;
var TWELVE_HOUR_FORMAT = /[h]{1}/;
var HM_FORMAT = /[hH]{1,2}:m{1,2}/;
var HMS_FORMAT = /[hH]{1,2}:m{1,2}:s{1,2}/;

exports.AM = AM;
exports.DEFAULT_FORMAT = DEFAULT_FORMAT;
exports.DEFAULT_STEPS = DEFAULT_STEPS;
exports.EMPTY_VALUE = EMPTY_VALUE;
exports.EPickerCols = EPickerCols;
exports.HMS_FORMAT = HMS_FORMAT;
exports.HM_FORMAT = HM_FORMAT;
exports.MERIDIEM_LIST = MERIDIEM_LIST;
exports.PM = PM;
exports.POST_MERIDIEM_FORMAT = POST_MERIDIEM_FORMAT;
exports.PRE_MERIDIEM_FORMAT = PRE_MERIDIEM_FORMAT;
exports.TIME_FORMAT = TIME_FORMAT;
exports.TIME_PICKER_EMPTY = TIME_PICKER_EMPTY;
exports.TWELVE_HOUR_FORMAT = TWELVE_HOUR_FORMAT;
