// tdesign-vue-next/cjs/sticky-tool/props.js
/**
 * tdesign v1.9.4
 * (c) 2024 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var props = {
  list: {
    type: Array,
    "default": function _default() {
      return [];
    }
  },
  offset: {
    type: Array
  },
  placement: {
    type: String,
    "default": "right-bottom",
    validator: function validator(val) {
      if (!val) return true;
      return ["right-top", "right-center", "right-bottom", "left-top", "left-center", "left-bottom"].includes(val);
    }
  },
  popupProps: {
    type: Object
  },
  shape: {
    type: String,
    "default": "square",
    validator: function validator(val) {
      if (!val) return true;
      return ["square", "round"].includes(val);
    }
  },
  type: {
    type: String,
    "default": "normal",
    validator: function validator(val) {
      if (!val) return true;
      return ["normal", "compact"].includes(val);
    }
  },
  width: {
    type: [String, Number]
  },
  onClick: Function,
  onHover: Function
};

exports["default"] = props;
