// tdesign-vue-next/lib/_chunks/dep-218bce82.js
/**
 * tdesign v1.9.4
 * (c) 2024 tdesign
 * @license MIT
 */

import { Button } from '../button/index.js';
import { Icon } from '../icon/index.js';
import { Link } from '../link/index.js';
import { Row, Col } from '../grid/index.js';
import { Aside, Layout, Header, Footer, Content } from '../layout/index.js';
import { Divider } from '../divider/index.js';
import { Space } from '../space/index.js';
import { Descriptions, DescriptionsItem } from '../descriptions/index.js';
import { Affix } from '../affix/index.js';
import { Anchor, AnchorItem, AnchorTarget } from '../anchor/index.js';
import { Breadcrumb, BreadcrumbItem } from '../breadcrumb/index.js';
import { Dropdown, DropdownItem, DropdownMenu } from '../dropdown/index.js';
import { Menu, HeadMenu, Submenu, MenuItem, MenuGroup } from '../menu/index.js';
import { Pagination, PaginationMini } from '../pagination/index.js';
import { Steps, StepItem } from '../steps/index.js';
import { Tabs, TabPanel } from '../tabs/index.js';
import { StickyItem, StickyTool } from '../sticky-tool/index.js';
import { BackTop } from '../back-top/index.js';
import { AutoComplete, HighlightOption } from '../auto-complete/index.js';
import { Cascader, CascaderPanel } from '../cascader/index.js';
import { Checkbox, CheckboxGroup } from '../checkbox/index.js';
import { DatePicker, DatePickerPanel, DateRangePicker, DateRangePickerPanel } from '../date-picker/index.js';
import { Form, FormItem } from '../form/index.js';
import { InputAdornment } from '../input-adornment/index.js';
import { Input, InputGroup } from '../input/index.js';
import { InputNumber } from '../input-number/index.js';
import { Radio, RadioGroup, RadioButton } from '../radio/index.js';
import { Select, Option, OptionGroup } from '../select/index.js';
import { Slider, SliderButton } from '../slider/index.js';
import { Switch } from '../switch/index.js';
import { TagInput } from '../tag-input/index.js';
import { SelectInput } from '../select-input/index.js';
import { Textarea } from '../textarea/index.js';
import { Transfer } from '../transfer/index.js';
import { TimePicker, TimeRangePicker, TimePickerPanel } from '../time-picker/index.js';
import { TreeSelect } from '../tree-select/index.js';
import { ColorPickerPanel, ColorPicker } from '../color-picker/index.js';
import { RangeInput, RangeInputPopup } from '../range-input/index.js';
import { Avatar, AvatarGroup } from '../avatar/index.js';
import { Badge } from '../badge/index.js';
import { Calendar } from '../calendar/index.js';
import { Card } from '../card/index.js';
import { Comment } from '../comment/index.js';
import { Image } from '../image/index.js';
import { ImageViewer } from '../image-viewer/index.js';
import { List, ListItem, ListItemMeta } from '../list/index.js';
import { Progress } from '../progress/index.js';
import { Skeleton } from '../skeleton/index.js';
import { Statistic } from '../statistic/index.js';
import { Swiper, SwiperItem } from '../swiper/index.js';
import { BaseTable, PrimaryTable, EnhancedTable, Table } from '../table/index.js';
import { Tag, CheckTag, CheckTagGroup } from '../tag/index.js';
import { Timeline, TimelineItem } from '../timeline/index.js';
import { Tooltip } from '../tooltip/index.js';
import { Tree } from '../tree/index.js';
import { Collapse, CollapsePanel } from '../collapse/index.js';
import { Watermark } from '../watermark/index.js';
import { Rate } from '../rate/index.js';
import { Alert } from '../alert/index.js';
import { Dialog } from '../dialog/index.js';
import { Drawer } from '../drawer/index.js';
import { Guide } from '../guide/index.js';
import { Loading } from '../loading/index.js';
import { Message } from '../message/index.js';
import { Notification } from '../notification/index.js';
import { Popconfirm } from '../popconfirm/index.js';
import { Popup } from '../popup/index.js';
import { Upload } from '../upload/index.js';
import { ConfigProvider } from '../config-provider/index.js';
import { DialogPlugin } from '../dialog/plugin.js';
import { LoadingPlugin } from '../loading/plugin.js';
import { vLoading } from '../loading/directive.js';
import { MessagePlugin } from '../message/plugin.js';
import NotificationPlugin from '../notification/plugin.js';

var components = /*#__PURE__*/Object.freeze({
	__proto__: null,
	Button: Button,
	Icon: Icon,
	Link: Link,
	Row: Row,
	Col: Col,
	Aside: Aside,
	Layout: Layout,
	Header: Header,
	Footer: Footer,
	Content: Content,
	Divider: Divider,
	Space: Space,
	Descriptions: Descriptions,
	DescriptionsItem: DescriptionsItem,
	Affix: Affix,
	Anchor: Anchor,
	AnchorItem: AnchorItem,
	AnchorTarget: AnchorTarget,
	Breadcrumb: Breadcrumb,
	BreadcrumbItem: BreadcrumbItem,
	Dropdown: Dropdown,
	DropdownItem: DropdownItem,
	DropdownMenu: DropdownMenu,
	Menu: Menu,
	HeadMenu: HeadMenu,
	Submenu: Submenu,
	MenuItem: MenuItem,
	MenuGroup: MenuGroup,
	Pagination: Pagination,
	PaginationMini: PaginationMini,
	Steps: Steps,
	StepItem: StepItem,
	Tabs: Tabs,
	TabPanel: TabPanel,
	StickyItem: StickyItem,
	StickyTool: StickyTool,
	BackTop: BackTop,
	AutoComplete: AutoComplete,
	HighlightOption: HighlightOption,
	Cascader: Cascader,
	CascaderPanel: CascaderPanel,
	Checkbox: Checkbox,
	CheckboxGroup: CheckboxGroup,
	DatePicker: DatePicker,
	DatePickerPanel: DatePickerPanel,
	DateRangePicker: DateRangePicker,
	DateRangePickerPanel: DateRangePickerPanel,
	Form: Form,
	FormItem: FormItem,
	InputAdornment: InputAdornment,
	Input: Input,
	InputGroup: InputGroup,
	InputNumber: InputNumber,
	Radio: Radio,
	RadioGroup: RadioGroup,
	RadioButton: RadioButton,
	Select: Select,
	Option: Option,
	OptionGroup: OptionGroup,
	Slider: Slider,
	SliderButton: SliderButton,
	Switch: Switch,
	TagInput: TagInput,
	SelectInput: SelectInput,
	Textarea: Textarea,
	Transfer: Transfer,
	TimePicker: TimePicker,
	TimeRangePicker: TimeRangePicker,
	TimePickerPanel: TimePickerPanel,
	TreeSelect: TreeSelect,
	ColorPickerPanel: ColorPickerPanel,
	ColorPicker: ColorPicker,
	RangeInput: RangeInput,
	RangeInputPopup: RangeInputPopup,
	Avatar: Avatar,
	AvatarGroup: AvatarGroup,
	Badge: Badge,
	Calendar: Calendar,
	Card: Card,
	Comment: Comment,
	Image: Image,
	ImageViewer: ImageViewer,
	List: List,
	ListItem: ListItem,
	ListItemMeta: ListItemMeta,
	Progress: Progress,
	Skeleton: Skeleton,
	Statistic: Statistic,
	Swiper: Swiper,
	SwiperItem: SwiperItem,
	BaseTable: BaseTable,
	PrimaryTable: PrimaryTable,
	EnhancedTable: EnhancedTable,
	Table: Table,
	Tag: Tag,
	CheckTag: CheckTag,
	CheckTagGroup: CheckTagGroup,
	Timeline: Timeline,
	TimelineItem: TimelineItem,
	Tooltip: Tooltip,
	Tree: Tree,
	Collapse: Collapse,
	CollapsePanel: CollapsePanel,
	Watermark: Watermark,
	Rate: Rate,
	Alert: Alert,
	DialogPlugin: DialogPlugin,
	Dialog: Dialog,
	Drawer: Drawer,
	Guide: Guide,
	LoadingPlugin: LoadingPlugin,
	LoadingDirective: vLoading,
	Loading: Loading,
	MessagePlugin: MessagePlugin,
	Message: Message,
	NotifyPlugin: NotificationPlugin,
	Notification: Notification,
	Popconfirm: Popconfirm,
	Popup: Popup,
	Upload: Upload,
	ConfigProvider: ConfigProvider
});

export { components as c };
