// tdesign-vue-next/cjs/_common/js/common.js
/**
 * tdesign v1.9.4
 * (c) 2024 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var EKeyboardDirection = /* @__PURE__ */function (EKeyboardDirection2) {
  EKeyboardDirection2[EKeyboardDirection2["left"] = 37] = "left";
  EKeyboardDirection2[EKeyboardDirection2["up"] = 38] = "up";
  EKeyboardDirection2[EKeyboardDirection2["right"] = 39] = "right";
  EKeyboardDirection2[EKeyboardDirection2["down"] = 40] = "down";
  return EKeyboardDirection2;
}(EKeyboardDirection || {});
var ARROW_DOWN_REG = /^ArrowDown$/i;
var ARROW_UP_REG = /^ArrowUp$/i;
var ESCAPE_REG = /^Escape$/i;
var SPACE_REG = /^Space$/i;
var ENTER_REG = /^Enter$/i;
var SHIFT_REG = /^(Shift|ShiftLeft|ShiftRight)$/i;
var CLEAR_REG = /^KeyC$/i;
var ALL_REG = /^(KeyA|KeyL)$/i;
var CHECKED_CODE_REG = /^(Enter|Space)$/i;

exports.ALL_REG = ALL_REG;
exports.ARROW_DOWN_REG = ARROW_DOWN_REG;
exports.ARROW_UP_REG = ARROW_UP_REG;
exports.CHECKED_CODE_REG = CHECKED_CODE_REG;
exports.CLEAR_REG = CLEAR_REG;
exports.EKeyboardDirection = EKeyboardDirection;
exports.ENTER_REG = ENTER_REG;
exports.ESCAPE_REG = ESCAPE_REG;
exports.SHIFT_REG = SHIFT_REG;
exports.SPACE_REG = SPACE_REG;
