// tdesign-vue-next/cjs/tree/props.js
/**
 * tdesign v1.9.4
 * (c) 2024 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var props = {
  activable: Boolean,
  activeMultiple: Boolean,
  actived: {
    type: Array,
    "default": void 0
  },
  defaultActived: {
    type: Array
  },
  allowFoldNodeOnFilter: Boolean,
  checkProps: {
    type: Object
  },
  checkStrictly: Boolean,
  checkable: Boolean,
  data: {
    type: Array,
    "default": function _default() {
      return [];
    }
  },
  disableCheck: {
    type: [Boolean, Function],
    "default": false
  },
  disabled: Boolean,
  draggable: Boolean,
  empty: {
    type: [String, Function],
    "default": ""
  },
  expandAll: Boolean,
  expandLevel: {
    type: Number,
    "default": 0
  },
  expandMutex: Boolean,
  expandOnClickNode: Boolean,
  expandParent: Boolean,
  expanded: {
    type: Array,
    "default": void 0
  },
  defaultExpanded: {
    type: Array,
    "default": function _default() {
      return [];
    }
  },
  filter: {
    type: Function
  },
  height: {
    type: [String, Number]
  },
  hover: Boolean,
  icon: {
    type: [Boolean, Function],
    "default": true
  },
  keys: {
    type: Object
  },
  label: {
    type: [String, Boolean, Function],
    "default": true
  },
  lazy: {
    type: Boolean,
    "default": true
  },
  line: {
    type: [Boolean, Function],
    "default": false
  },
  load: {
    type: Function
  },
  maxHeight: {
    type: [String, Number]
  },
  operations: {
    type: Function
  },
  scroll: {
    type: Object
  },
  transition: {
    type: Boolean,
    "default": true
  },
  value: {
    type: Array,
    "default": void 0
  },
  modelValue: {
    type: Array,
    "default": void 0
  },
  defaultValue: {
    type: Array,
    "default": function _default() {
      return [];
    }
  },
  valueMode: {
    type: String,
    "default": "onlyLeaf",
    validator: function validator(val) {
      if (!val) return true;
      return ["onlyLeaf", "parentFirst", "all"].includes(val);
    }
  },
  onActive: Function,
  onChange: Function,
  onClick: Function,
  onDragEnd: Function,
  onDragLeave: Function,
  onDragOver: Function,
  onDragStart: Function,
  onDrop: Function,
  onExpand: Function,
  onLoad: Function,
  onScroll: Function
};

exports["default"] = props;
