// tdesign-vue-next/cjs/textarea/props.js
/**
 * tdesign v1.9.4
 * (c) 2024 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var props = {
  allowInputOverMax: {
    type: Boolean,
    "default": false
  },
  autofocus: Boolean,
  autosize: {
    type: [Boolean, Object],
    "default": false
  },
  disabled: Boolean,
  maxcharacter: {
    type: Number
  },
  maxlength: {
    type: [Number, String]
  },
  name: {
    type: String,
    "default": ""
  },
  placeholder: {
    type: String,
    "default": void 0
  },
  readonly: Boolean,
  status: {
    type: String,
    validator: function validator(val) {
      return ["default", "success", "warning", "error"].includes(val);
    }
  },
  tips: {
    type: [String, Function]
  },
  value: {
    type: [String, Number]
  },
  modelValue: {
    type: [String, Number]
  },
  defaultValue: {
    type: [String, Number]
  },
  onBlur: Function,
  onChange: Function,
  onFocus: Function,
  onKeydown: Function,
  onKeypress: Function,
  onKeyup: Function,
  onValidate: Function
};

exports["default"] = props;
