// tdesign-vue-next/cjs/image-viewer/props.js
/**
 * tdesign v1.9.4
 * (c) 2024 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var props = {
  closeBtn: {
    type: [Boolean, Function],
    "default": true
  },
  closeOnEscKeydown: {
    type: Boolean,
    "default": true
  },
  closeOnOverlay: Boolean,
  draggable: {
    type: Boolean,
    "default": void 0
  },
  imageScale: {
    type: Object
  },
  images: {
    type: Array,
    "default": function _default() {
      return [];
    }
  },
  index: {
    type: Number,
    "default": void 0
  },
  defaultIndex: {
    type: Number,
    "default": 0
  },
  mode: {
    type: String,
    "default": "modal",
    validator: function validator(val) {
      if (!val) return true;
      return ["modal", "modeless"].includes(val);
    }
  },
  navigationArrow: {
    type: [Boolean, Function],
    "default": true
  },
  showOverlay: {
    type: Boolean,
    "default": void 0
  },
  title: {
    type: [String, Function]
  },
  trigger: {
    type: [String, Function]
  },
  viewerScale: {
    type: Object
  },
  visible: {
    type: Boolean,
    "default": void 0
  },
  modelValue: {
    type: Boolean,
    "default": void 0
  },
  defaultVisible: Boolean,
  zIndex: {
    type: Number
  },
  onClose: Function,
  onIndexChange: Function
};

exports["default"] = props;
