// tdesign-vue-next/cjs/table/base-table-props.js
/**
 * tdesign v1.9.4
 * (c) 2024 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var baseTableProps = {
  activeRowKeys: {
    type: Array,
    "default": void 0
  },
  defaultActiveRowKeys: {
    type: Array,
    "default": function _default() {
      return [];
    }
  },
  activeRowType: {
    type: String,
    "default": ""
  },
  allowResizeColumnWidth: {
    type: Boolean,
    "default": void 0
  },
  attach: {
    type: [String, Function]
  },
  bordered: Boolean,
  bottomContent: {
    type: [String, Function]
  },
  cellEmptyContent: {
    type: [String, Function]
  },
  columns: {
    type: Array,
    "default": function _default() {
      return [];
    }
  },
  data: {
    type: Array,
    "default": function _default() {
      return [];
    }
  },
  disableDataPage: Boolean,
  disableSpaceInactiveRow: {
    type: Boolean,
    "default": void 0
  },
  empty: {
    type: [String, Function],
    "default": ""
  },
  firstFullRow: {
    type: [String, Function]
  },
  fixedRows: {
    type: Array
  },
  footData: {
    type: Array,
    "default": function _default() {
      return [];
    }
  },
  footerAffixProps: {
    type: Object
  },
  footerAffixedBottom: {
    type: [Boolean, Object],
    "default": false
  },
  footerSummary: {
    type: [String, Function]
  },
  headerAffixProps: {
    type: Object
  },
  headerAffixedTop: {
    type: [Boolean, Object],
    "default": false
  },
  height: {
    type: [String, Number]
  },
  horizontalScrollAffixedBottom: {
    type: [Boolean, Object]
  },
  hover: Boolean,
  keyboardRowHover: {
    type: Boolean,
    "default": true
  },
  lastFullRow: {
    type: [String, Function]
  },
  lazyLoad: Boolean,
  loading: {
    type: [Boolean, Function],
    "default": void 0
  },
  loadingProps: {
    type: Object
  },
  locale: {
    type: Object
  },
  maxHeight: {
    type: [String, Number]
  },
  pagination: {
    type: Object
  },
  paginationAffixedBottom: {
    type: [Boolean, Object]
  },
  resizable: Boolean,
  rowAttributes: {
    type: [Object, Array, Function]
  },
  rowClassName: {
    type: [String, Object, Array, Function]
  },
  rowKey: {
    type: String,
    "default": "id",
    required: true
  },
  rowspanAndColspan: {
    type: Function
  },
  rowspanAndColspanInFooter: {
    type: Function
  },
  scroll: {
    type: Object
  },
  showHeader: {
    type: Boolean,
    "default": true
  },
  size: {
    type: String,
    validator: function validator(val) {
      if (!val) return true;
      return ["small", "medium", "large"].includes(val);
    }
  },
  stripe: Boolean,
  tableContentWidth: {
    type: String,
    "default": ""
  },
  tableLayout: {
    type: String,
    "default": "fixed",
    validator: function validator(val) {
      if (!val) return true;
      return ["auto", "fixed"].includes(val);
    }
  },
  topContent: {
    type: [String, Function]
  },
  verticalAlign: {
    type: String,
    "default": "middle",
    validator: function validator(val) {
      if (!val) return true;
      return ["top", "middle", "bottom"].includes(val);
    }
  },
  onActiveChange: Function,
  onActiveRowAction: Function,
  onCellClick: Function,
  onColumnResizeChange: Function,
  onPageChange: Function,
  onRowClick: Function,
  onRowDblclick: Function,
  onRowMousedown: Function,
  onRowMouseenter: Function,
  onRowMouseleave: Function,
  onRowMouseover: Function,
  onRowMouseup: Function,
  onScroll: Function,
  onScrollX: Function,
  onScrollY: Function
};

exports["default"] = baseTableProps;
