// tdesign-vue-next/cjs/range-input/props.js
/**
 * tdesign v1.9.4
 * (c) 2024 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var props = {
  activeIndex: {
    type: Number
  },
  clearable: Boolean,
  disabled: Boolean,
  format: {
    type: [Array, Function]
  },
  inputProps: {
    type: [Object, Array]
  },
  label: {
    type: [String, Function]
  },
  placeholder: {
    type: [String, Array]
  },
  prefixIcon: {
    type: Function
  },
  readonly: Boolean,
  separator: {
    type: [String, Function],
    "default": "-"
  },
  showClearIconOnEmpty: Boolean,
  size: {
    type: String,
    "default": "medium",
    validator: function validator(val) {
      if (!val) return true;
      return ["small", "medium", "large"].includes(val);
    }
  },
  status: {
    type: String,
    "default": "default",
    validator: function validator(val) {
      if (!val) return true;
      return ["default", "success", "warning", "error"].includes(val);
    }
  },
  suffix: {
    type: [String, Function]
  },
  suffixIcon: {
    type: Function
  },
  tips: {
    type: [String, Function]
  },
  value: {
    type: Array,
    "default": void 0
  },
  modelValue: {
    type: Array,
    "default": void 0
  },
  defaultValue: {
    type: Array,
    "default": function _default() {
      return [];
    }
  },
  onBlur: Function,
  onChange: Function,
  onClear: Function,
  onClick: Function,
  onEnter: Function,
  onFocus: Function,
  onMouseenter: Function,
  onMouseleave: Function
};

exports["default"] = props;
